Microsoft.Toolkit/Observables/ObservableGroup.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+
+namespace Microsoft.Toolkit.Observables.Collections
+{
+    /// <summary>
+    /// An observable group.
+    /// It associates a <see cref="Key"/> to an <see cref="ObservableCollection{T}"/>.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the group key.</typeparam>
+    /// <typeparam name="TValue">The type of the items in the collection.</typeparam>
+    public sealed class ObservableGroup<TKey, TValue> : ObservableCollection<TValue>, IGrouping<TKey, TValue>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ObservableGroup{TKey, TValue}"/> class.
+        /// </summary>
+        /// <param name="key">The key for the group.</param>
+        public ObservableGroup(TKey key)
+        {
+            Key = key;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ObservableGroup{TKey, TValue}"/> class.
+        /// </summary>
+        /// <param name="grouping">The grouping to fill the group.</param>
+        public ObservableGroup(IGrouping<TKey, TValue> grouping)
+            : base(grouping)
+        {
+            Key = grouping.Key;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ObservableGroup{TKey, TValue}"/> class.
+        /// </summary>
+        /// <param name="key">The key for the group.</param>
+        /// <param name="collection">The initial collection of data to add to the group.</param>
+        public ObservableGroup(TKey key, IEnumerable<TValue> collection)
+            : base(collection)
+        {
+            Key = key;
+        }
+
+        /// <summary>
+        /// Gets the key of the group.
+        /// </summary>
+        public TKey Key { get; }
```

Comment:
This unfortunately can't be used from XAML, as the UWP-XAML syntax doesn't support generic types in the `x:DataType` definition. A workaround I'm using in my own app (which is virtually identical to your implementation here, other than this detail), is to introduce this interface for the `ObservableGroup<TKey, TValue>` type (as well as the readonly version, in this PR):

```csharp
/// <summary>
/// An <see langword="interface"/> for a grouped collection of items.
/// </summary>
public interface IObservableGroup
{
    /// <summary>
    /// Gets the key for the current collection, as an <see cref="object"/>.
    /// </summary>
    object Key { get; }

    /// <summary>
    /// Gets the number of items currently in the grouped collection.
    /// </summary>
    int Count { get; }
}
```

The `Count` property is handy from XAML and is automatically implemented by the underlying `ObservableCollection<TValue>` type, while the `Key` property should be implemented explicitly, like so:

```csharp
/// <inheritdoc/>
object IObservableGroup.Key => Key;
```

This makes it possible to do something like this in XAML:

```xml
xmlns:system="using:System"

<DataTemplate
    x:Key="SomeGroupTemplate"
    x:DataType="objectModel:IObservableGroup">
    <StackPanel>
        <TextBlock Text="{x:Bind Key.ToString()}"/>
        <TextBlock Text="{x:Bind system:String.Format('{0} item(s)', Count)}"/>
    </StackPanel>
</DataTemplate>
```

This has no overhead if not used, but I'd argue it's an extremely valuable feature to have 😄

Reply:
Nice trick ! I will add it in this PR.

Reply:
Awesome, thanks! 😄🎉

Reply:
I've added `IReadOnlyObservableGroup` and updated the sample code.

Reply:
I've updated the XAML code to display the count value in the last commit.